module/zfs/zfs_log.c

```diff
@@ -525,6 +526,117 @@ zfs_log_rename(zilog_t *zilog, dmu_tx_t *tx, uint64_t txtype, znode_t *sdzp,
+ *     RENAME_WHITEOUT: creates a whiteout inode in place of renamed file as
+ *                      an atomic operation
+ *
+ * Ideally, these operations should be represented as new ZFS Intent Log
+ * txtypes, which would mandate a new ZFS feature flag due to the on-disk
+ * format change.  One would then use spa_feature_incr/decr functions to
+ * indicate that the on-disk log contains these new txtypes.  However, these
+ * functions are only supposed to be called from the txg syncing context.
+ *
+ * This means that we would need to force out an in-progress txg to disk and
+ * start a new one before writing any ZIL records.  This would ensure that
+ * previous versions of ZFS which do not support these log txtypes would
+ * never encounter them during ZIL replay.  Doing this would hurt performance.
+ *
+ * Alternatively, we could just activate the feature on a pool when these
+ * renameat2 flags get first used and leave it at that.  This would render
+ * the pool read-only importable on implementations without the new feature
+ * flag, even when no new txtypes were present on-disk.  This could be almost
+ * all of the time, so it'd be a shame to render the pool read-only on
+ * non-Linux platforms.
+ *
+ * Instead, we choose to rely on the fact that the ZIL is replayed in single-
+ * threaded mode before the dataset is mounted.  This means we can represent
+ * the otherwise atomic operations as a series of plain good old txtypes
+ * known to all current OpenZFS implementations. To do that, we use the
+ * following functions (at least until more platforms implement renameat2).
+ *
+ *     zfs_log_rename_exchange
+ *     zfs_log_rename_whiteout
+ */
+
+void
+zfs_log_rename_exchange(zilog_t *zilog, dmu_tx_t *tx, uint64_t txtype,
+    znode_t *sdzp, const char *sname, znode_t *tdzp, const char *dname,
+    znode_t *szp)
+{
+#ifndef __FreeBSD__
+	zfs_dirlock_t *tmpdl;
+#endif
+	znode_t *tmpzp = NULL;
+	char *tmpname;
+	int retries = 0;
+	int pos = 0;
+	int err;
+
+	tmpname = kmem_alloc(MAXPATHLEN, KM_SLEEP);
+
+	/*
+	 * To represent atomic rename with old non-atomic operations, we need
+	 * a temporary new name; so we try picking a name until we succeed,
+	 * then we get a dirent lock for that temp name until the final itx
+	 * gets queued
+	 */
+retry:
+	retries++;
+	pos = snprintf(tmpname, MAXPATHLEN, "%s.zfs_renameat2_emul_", dname);
+
+	for (int i = 0; i < 16; i++) {
+		int r = 0xFF;
+		random_get_pseudo_bytes((void *)&r, 1);
+		pos += snprintf(tmpname+pos, MAXPATHLEN-pos, "%02x", r);
+	}
+
+#ifdef __FreeBSD__
+	/*
+	 * NOTE: This code is never called on FreeBSD because they don't have
+	 *       RENAME_EXCHANGE, so it is possible this code is missing
+	 *       correct locking (though as far as I can tell, that is not the
+	 *       case).
+	 */
+	err = zfs_dirent_lookup(tdzp, tmpname, &tmpzp, ZNEW);
+#else
+	err = zfs_dirent_lock(&tmpdl, tdzp, tmpname, &tmpzp, ZNEW, NULL, NULL);
+#endif
+
+	VERIFY3U(retries, <, 10);
```

Comment:
Can we constrain this to VERIFY that only `0` and `EEXISTS` are value return codes?  In the `EEXISTS` case we should be able to retry indefinitely (though of course even a single collisions would be exceedingly unlikely).

```suggestion
	VERIFY(err == 0 || err == EEXISTS);
```

Reply:
We can, my only worry is that it is still hypothetically possible for someone to force the ZIL replay into an almost-infinite-loop (though in fairness `255^16` is a very large number of files -- 319626579315078487616775634918212890625 to be exact). Maybe we should bump the size up to 32 to really cover our bases?

Reply:
For reference the maximum size of a TXG is 4G, which means at most we're looking at million of renames which could randomly collide (unless the directory is pre-populated with names which match this pattern).  Increasing it to 32 bytes wouldn't really cost anything, that would be 256 bits or the same size as the ZFS checksums where we effectively never see collisions over the entire pool.  So it's hard to imagine that wouldn't be fine unless the PRNG was seriously broken.

Of course, if we did just go ahead and add the next TX records we wouldn't need this bit of ugliness at all since we could do it atomically with a single record.